docs/BLOCK_DEVICE_LOCKING.md

```diff
@@ -98,3 +98,145 @@ Summarizing: it is recommended to take `LOCK_EX` BSD file locks when
+Error happened: -1
+```
+
+If the version of `systemd` is not satisfied, download and compile the satisfied version of `systemd` source code. Assuming the code of `systemd` is placed in `/path/to/systemd/` and the targets are placed in `/path/to/systemd/build/`, run the following commands:
+```
+# gcc main.c -L/path/to/systemd/build/ -lsystemd -I/path/to/systemd/src/ -o main
+
+# LD_LIBRARY_PATH=/path/to/systemd/build:${LD_LIBRARY_PATH} ./main /dev/sda1
+Successfully get the lock: /dev/sda
+
+# LD_LIBRARY_PATH=/path/to/systemd/build:${LD_LIBRARY_PATH} flock -x /dev/sda ./main /dev/sda1
+Failed to get the lock: /dev/sda
+Error happened: -1
+```
+
+## Code
+```
+/* SPDX-License-Identifier: MIT-0 */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/file.h>
+#include <systemd/sd-device.h>
+#include <unistd.h>
+
+/**
+ * lock_whole_disk_from_devname
+ * @devname: devname of a block device, e.g., /dev/sda or /dev/sda1
+ * @open_mode: the mode to open the device, e.g., O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY
+ * @lock_mode: the mode to add flock, e.g., LOCK_EX|LOCK_NB
+ * @ret_fd: return the file description if successfully locked
+ *
+ * given the devname of a block device, lock the whole disk
+ *
+ * Returns: 0 if add lock successfully
+ **/
+int lock_whole_disk_from_devname(const char *devname, int open_mode, int lock_mode, int *ret_fd) {
+    sd_device *dev, *whole_dev;
+    const char *whole_disk_devname, *subsystem, *devtype;
+    int r, fd = -1;
+
+    if (ret_fd == NULL)
+        return -EINVAL;
+
+    // create a sd_device instance from devname
+    r = sd_device_new_from_devname(&dev, devname);
+    if (r < 0)
+        return r;
+
+    // if the subsystem of dev is block, but its devtype is not disk, find its parent
+    r = sd_device_get_subsystem(dev, &subsystem);
+    if (r < 0)
+        goto finalize;
+    if (strcmp(subsystem, "block") != 0) {
+        r = -EINVAL;
+        goto finalize;
+    }
+
+    r = sd_device_get_devtype(dev, &devtype);
+    if (r < 0)
+        goto finalize;
+    if (strcmp(devtype, "disk") == 0)
+        whole_dev = dev;
+    else {
+        r = sd_device_get_parent_with_subsystem_devtype(dev, "block", "disk", &whole_dev);
+        if (r < 0)
+            goto finalize;
+    }
+
+    // open the whole disk device node
+    fd = sd_device_open(whole_dev, open_mode);
+    if (fd < 0) {
+        r = fd;
+        goto finalize;
+    }
+
+    // get the whole disk devname
+    r = sd_device_get_devname(whole_dev, &whole_disk_devname);
+    if (r < 0)
+        goto finalize;
+
+    // add BSD lock on the whole disk device node
+    if (flock(fd, lock_mode) < 0) {
+        r = -errno;
+        printf("Failed to get the lock: %s\n", whole_disk_devname);
+        goto finalize;
+    }
+
+    printf("Successfully get the lock: %s\n", whole_disk_devname);
+    *ret_fd = fd;
+    fd = -1;
+    r = 0;
+
+finalize:
+    if (fd >= 0)
+        close(fd);
+
+    sd_device_unref(dev);
```

Comment:
`sd_device_unrefp` has been exported. Manual cleanups like this is something of an antipattern. Wouldn't it be better to use `__attribute__((cleanup(sd_device_unrefp))) sd_device *dev = NULL;` instead?